File: php/src/Intent/ChargeRequest.php
<?php

declare(strict_types=1);

namespace SolanaMpp\Intent;

use InvalidArgumentException;
use SolanaMpp\Core\Json;

/**
 * Represents the MPP charge intent request embedded in a challenge.
 */
final class ChargeRequest
{
    /**
     * Create an MPP charge request using base-unit integer amounts.
     *
     * @param array<string, mixed>|null $methodDetails
     */
    public function __construct(
        public readonly string $amount,
        public readonly string $currency,
        public readonly string $recipient = '',
        public readonly string $description = '',
        public readonly string $externalId = '',
        public readonly ?array $methodDetails = null,
    ) {
        self::assertBaseUnits($amount, 'amount');
        if ($currency === '') {
            throw new InvalidArgumentException('currency is required');
        }
    }

    /**
     * Convert the charge request to the Payment challenge request object.
     *
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        $value = [
            'amount' => $this->amount,
            'currency' => $this->currency,
        ];
        if ($this->recipient !== '') {
            $value['recipient'] = $this->recipient;
        }
        if ($this->description !== '') {
            $value['description'] = $this->description;
        }
        if ($this->externalId !== '') {
            $value['externalId'] = $this->externalId;
        }
        if ($this->methodDetails !== null) {
            $value['methodDetails'] = $this->methodDetails;
        }

        return $value;
    }

    /**
     * Decode a Payment challenge request object.
     *
     * @param array<string, mixed> $value
     */
    public static function fromArray(array $value): self
    {
        $methodDetails = $value['methodDetails'] ?? null;
        if ($methodDetails !== null && !is_array($methodDetails)) {
            throw new InvalidArgumentException('methodDetails must be an object');
        }

        return new self(
            amount: Json::optionalString($value['amount'] ?? null, 'amount'),
            currency: Json::optionalString($value['currency'] ?? null, 'currency'),
            recipient: Json::optionalString($value['recipient'] ?? null, 'recipient'),
            description: Json::optionalString($value['description'] ?? null, 'description'),
            externalId: Json::optionalString($value['externalId'] ?? null, 'externalId'),
            methodDetails: is_array($methodDetails) ? Json::object($methodDetails, 'methodDetails') : null,
        );
    }

    private static function assertBaseUnits(string $value, string $field): void
    {
        if (
            $value === '' ||
            !ctype_digit($value) ||
            ltrim($value, '0') === '' ||
            (strlen($value) > 1 && str_starts_with($value, '0'))
        ) {
            throw new InvalidArgumentException(sprintf('%s must be a positive base-unit integer string', $field));
        }
    }
}

File: ruby/lib/mpp/intent/charge_request.rb
# frozen_string_literal: true

module Mpp
  module Intent
    # MPP charge request wire object.
    class ChargeRequest
      attr_reader :amount, :currency, :recipient, :description, :external_id, :method_details

      def initialize(amount:, currency:, recipient: nil, description: nil, external_id: nil, method_details: nil)
        raise ArgumentError, "amount must be a positive base-unit integer string" unless amount.to_s.match?(/\A[1-9][0-9]*\z/)
        raise ArgumentError, "currency is required" if currency.to_s.empty?
        raise ArgumentError, "methodDetails must be a Hash" unless method_details.nil? || method_details.is_a?(Hash)

        @amount = amount.to_s
        @currency = currency.to_s
        @recipient = recipient
        @description = description
        @external_id = external_id
        @method_details = method_details || {}
      end

      # Build a charge request from decoded wire JSON.
      def self.from_h(value)
        raise ArgumentError, "charge request must be an object" unless value.is_a?(Hash)

        new(
          amount: value.fetch("amount"),
          currency: value.fetch("currency"),
          recipient: value["recipient"],
          description: value["description"],
          external_id: value["externalId"],
          method_details: value["methodDetails"]
        )
      end

      # Convert a display amount to base units.
      def self.parse_units(amount, decimals)
        raw = amount.to_s
        raise ArgumentError, "invalid amount" unless raw.match?(/\A[0-9]+(\.[0-9]+)?\z/)

        whole, frac = raw.split(".", 2)
        frac ||= ""
        raise ArgumentError, "too many decimal places" if frac.length > decimals

        (whole + frac.ljust(decimals, "0")).sub(/\A0+(?=\d)/, "")
      end

      # Serialize to the camelCase wire object.
      def to_h
        {
          "amount" => amount,
          "currency" => currency,
          "recipient" => recipient,
          "description" => description,
          "externalId" => external_id,
          "methodDetails" => method_details.empty? ? nil : method_details
        }.compact
      end

      # Parse the base-unit amount as an Integer.
      def amount_i
        Integer(amount, 10)
      rescue ArgumentError
        raise ArgumentError, "invalid amount: #{amount}"
      end
    end
  end
end
